### DISCUSSION TITLE
Docker installation for noobs

### DISCUSSION THREAD

**Discussion by 4wsa** [nioc/xmpp-web#125] [Q&A] [answered]
Hi, 

Sorry if this is the wrong place to ask.

So I managed to figure out how to use a VPS and I managed to get Snikket running despite still having no clue what a Docker does _(I think it's some sort of sandbox? Or fancy zip file? No clue. 🤷‍♂️)_ but I'm trying to install xmpp-web on my VPS and I'm really not sure what commands to input.

The [instructions](https://github.com/nioc/xmpp-web?tab=readme-ov-file#docker-image) say 
`docker run -it -p 80:80 --rm \
-e XMPP_WS=https://domain-xmpp.ltd:5281/xmpp-websocket \
-e APP_DEFAULT_DOMAIN=domain-xmpp.ltd \
--name xmpp-web-1 nioc/xmpp-web`

and they also suggest a `docker-compose.yml` file (which I remember from Snikket so I created said file in `/etc/xmpp-web/`) but unfortunately I really have no idea what to do next. None of the commands above worked, and I also tried `docker pull nioc/xmpp-web` hoping it would work like it did with Snikket.

Is there a list of commands for total idiots to copy and paste into their SSH terminal to get xmpp-web working?

**Comment by nioc** [comment]
Hello @4wsa , you can find commands for Docker installation here:
https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository 

Docker can be seen as a sandbox, it allows to deploy easily app from "container" pulled from docker hub.

You can also install a webserver and configure it but it needs more steps.

**Reply by 4wsa** [reply]
> Hello @4wsa , you can find commands for Docker installation here: https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository
> 
> Docker can be seen as a sandbox, it allows to deploy easily app from "container" pulled from docker hub.
> 
> You can also install a webserver and configure it but it needs more steps.

Thank you for the reply, but I actually already have Docker installed and I also have Snikket installed and running. I'm just trying to find the idiot-proof commands to install xmpp-web.

Basically the chain of commands that end up in me having xmpp-web installed and booted. 😁

**Reply by nioc** [reply]
If you already have a docker engine running, just type the command as mentioned in the doc:
```
docker run -it -p 80:80 --rm \
-e XMPP_WS=https://domain-xmpp.ltd:5281/xmpp-websocket \
-e APP_DEFAULT_DOMAIN=domain-xmpp.ltd \
--name xmpp-web-1 nioc/xmpp-web
```
Set/add the environment variables `-e XXX` according to the [config](https://github.com/nioc/xmpp-web?tab=readme-ov-file#configuration) you need.

**Reply by 4wsa** [reply]
> If you already have a docker engine running, just type the command as mentioned in the doc:
> 
> ```
> docker run -it -p 80:80 --rm \
> -e XMPP_WS=https://domain-xmpp.ltd:5281/xmpp-websocket \
> -e APP_DEFAULT_DOMAIN=domain-xmpp.ltd \
> --name xmpp-web-1 nioc/xmpp-web
> ```
> 
> Set/add the environment variables `-e XXX` according to the [config](https://github.com/nioc/xmpp-web?tab=readme-ov-file#configuration) you need.

Right, that's where I'm having some trouble. I changed the command to this:

```
docker run -it -p 80:80 --rm \
-e XMPP_WS=https://**MYDOMAIN**:5281/xmpp-websocket \
-e APP_DEFAULT_DOMAIN=**MYDOMAIN** \
--name xmpp-web-1 nioc/xmpp-web
```

I also opened port 5281 with ufw.

This was the output.

`Unable to find image 'docker:latest' locally
latest: Pulling from library/docker
ec99f8b99825: Pull complete
827ba70d8228: Pull complete
4f4fb700ef54: Pull complete
b40059883d04: Pull complete
a0eef4938cb6: Pull complete
d4d3a0ac7415: Pull complete
90fc9b45b17b: Pull complete
5fcf5ea44c05: Pull complete
219e595ad0a8: Pull complete
2986c95c7e5a: Pull complete
921c0a0c766d: Pull complete
38376a055077: Pull complete
5791e1a08695: Pull complete
03a765629918: Pull complete
b7c748af66c9: Pull complete
17d6e0e26857: Pull complete
Digest: sha256:66776493766db8133f3d892b9c647551cd9bc5b8a948c57177d4ffbd4b33681e
Status: Downloaded newer image for docker:latest
docker: Error response from daemon: driver failed programming external connectivity on endpoint keen_bose (3267b75a8bced7da4ee42b4ba7e05b40146f4c2b449a65642f385e7d0e5b16bf): Error starting userland proxy: listen tcp4 0.0.0.0:80: bind: address already in use.
`

Afterwards I changed it to 81:81 because I thought maybe the port is in use by Snikket's http server but it ends up just hanging on a different output.

`/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf
10-listen-on-ipv6-by-default.sh: info: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
/docker-entrypoint.sh: Sourcing /docker-entrypoint.d/15-local-resolvers.envsh
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/default.conf.template to /etc/nginx/conf.d/default.conf
/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
/docker-entrypoint.sh: Launching /docker-entrypoint.d/998-update-local.js.sh
update local.js with environment variables... done
/docker-entrypoint.sh: Configuration complete; ready for start up
2024/06/22 12:54:25 [notice] 1#1: using the "epoll" event method
2024/06/22 12:54:25 [notice] 1#1: nginx/1.26.0
2024/06/22 12:54:25 [notice] 1#1: built by gcc 13.2.1 20231014 (Alpine 13.2.1_git20231014)
2024/06/22 12:54:25 [notice] 1#1: OS: Linux 4.19.0
2024/06/22 12:54:25 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2024/06/22 12:54:25 [notice] 1#1: start worker processes
2024/06/22 12:54:25 [notice] 1#1: start worker process 39
`

**Reply by nioc** [reply, marked as answer]
If port 80 of your host is already used, you can use another port with `-p 81:80` (`<host port>:<container port>`)
```
docker run -it -p 81:80 --rm \
-e XMPP_WS=https://**MYDOMAIN**:5281/xmpp-websocket \
-e APP_DEFAULT_DOMAIN=**MYDOMAIN** \
--name xmpp-web-1 nioc/xmpp-web
```

The open http://yourvps:81/

If you want to use https, I suggest using a proxy like haproxy or nginx with let's encrypt.

**Reply by 4wsa** [reply]
> If port 80 of your host is already used, you can use another port with `-p 81:80` (`<host port>:<container port>`)
> 
> ```
> docker run -it -p 81:80 --rm \
> -e XMPP_WS=https://**MYDOMAIN**:5281/xmpp-websocket \
> -e APP_DEFAULT_DOMAIN=**MYDOMAIN** \
> --name xmpp-web-1 nioc/xmpp-web
> ```
> 
> The open http://yourvps:81/
> 
> If you want to use https, I suggest using a proxy like haproxy or nginx with let's encrypt.

Thanks! That command worked and the login screen appears when I go to the correct URL. I don't care about setting up https (for now) so I just changed the command to this:
`docker run -it -p 81:80 --rm \
-e XMPP_WS=http://MYDOMAIN:5281/xmpp-websocket \
-e APP_DEFAULT_DOMAIN=**MYDOMAIN** \
--name xmpp-web-1 nioc/xmpp-web`

Unfortunately it's unable to login (`Server unreachable`) but that's beyond the scope of this thread. I wanted to change config.js to try to get it working but this is way beyond my skillset. I thought it would just be opening config.js but it seems far more complicated than that. https://stackoverflow.com/questions/48059540/how-to-change-docker-config-of-an-already-running-container

I really appreciate your help but I think I'm going to drop trying to get this tool running and find a simpler solution. 👍
